### ISSUE TITLE
Live Directive

### ISSUE THREAD

**Issue by ivelten** [fsprojects/FSharp.Data.GraphQL#165]
At the moment, we don't have support for [live directive](https://dev-blog.apollodata.com/new-features-in-graphql-batch-defer-stream-live-and-subscribe-7585d0c28b07).

There are many possible implementations for this, with two of them being initially suggested by the team:

**1. Add the possibility of defining live field definitions.**
The idea for this is to let the user define which fields can be queried with live directive at the moment of the schema definition. A live field definition would work almost like a normal field definition, except that if the type of the field is `'T`, it would require an `IObservable<'T>` as output on the resolver function. There is a WIP Pull Request currently open with this approach [here](https://github.com/fsprojects/FSharp.Data.GraphQL/pull/163). On the sample project, the implementation of a live field would be like:

```
let buildDateTime (dt : System.DateTime) =
    { Day = dt.Day
      Month = dt.Month
      Year = dt.Year
      Hour = dt.Hour
      Minute = dt.Minute
      Second = dt.Second }

let dateTime = 
    let span = System.TimeSpan.FromSeconds(float 1)
    Observable.Interval(span).Select(fun _ -> buildDateTime System.DateTime.Now)

let DateTimeType : ObjectDef<DateTime> =
    Define.Object<DateTime>(...)

let Query =
    Define.Object<Root>(
        name = "Query",
        fields = [
            Define.Field("time", Live DateTimeType, "Gets server's date", fun _ _ -> dateTime)])
```

**2. Using a live query subscription provider.**
We think this approach is better than the first one, because there's no need to do heavy field definition for it to work. The idea consists in having an `ILiveFieldSubscriptionProvider` in the `ISchema`. This provider will hold subscriptions for each field that the user wants to be able to go live:

```
type ILiveFieldSubscription =
    interface
        abstract member Identity : obj -> obj
        abstract member TypeName : string
        abstract member FieldName : string
    end

and ILiveFieldSubscription<'Object, 'Identity> =
    interface
        inherit ILiveFieldSubscription
        abstract member Identity : 'Object -> 'Identity
    end

and ILiveFieldSubscriptionProvider =
    interface
        abstract member IsRegistered : string -> string -> bool
        abstract member Register : ILiveFieldSubscription -> unit
        abstract member TryFindSubscription : string -> string -> ILiveFieldSubscription option
        abstract member Add : obj -> string -> string -> IObservable<obj>
        abstract member Publish<'T> : string -> string -> 'T -> unit
    end
```

A `ILiveFieldSubscription` will hold a function that given the object of the definition, returns the value that represents it's identity. When the object with the specified identity is updated, the new value is sent to all subscribers.

Finally, `ILiveFieldSubscriptionProvider` will push updates to observers when `Publish<'T>` is called, passing the type name, the field name, and the actual value of the object. To subscribe to a field, we call the Add function, passing the identity value of it, followed by the type name and the field name. Most of the time, this won't be needed, because it will be used internally by `Executor` class will to return live queries in a Deferred `GQLResponse`.

This approach also can come with a middleware, for example, that creates subscriptions for all fields that have a property named "Id". This would ease making an entire schema being able to go live, needing only to manage `Push<'T>` calls when values get updated.

There is also an open WIP Pull Request for this approach [here](https://github.com/fsprojects/FSharp.Data.GraphQL/pull/166).

**Comment by Horusiath** [comment]
Is it possible to generalize cases of deferred and live queries into a single abstraction? It's easy to multiply interfaces, but I'm trying to find out the case for making it a single composable contract.

Btw. are live queries part of the GraphQL standard by now?

**Comment by ivelten** [comment]
@Horusiath, I'm not sure if that's what you mean, but, actually, `defer`, `stream `and `live `queries are being returned into the `Deferred` case of `GQLResponse`. We thought that it could be returned there, since it works in the exact same way for the client (it only needs to subscribe on the values). There's no way to know, only in the response, which kind of result is being returned at the moment, though. I thought that we could change the name of the case at some point, but that would be a breaking change.

If you are talking about the Provider interfaces though, I didn't give that a thought, mainly because deferred and streamed queries don't need, at the moment, anything else to work (they just send the same response laters, in a socket).

From what I know, only `include `and `skip` directives are mandatory on server implementations. But defer, stream, export and live are being used by the Facebook for a while now, and because of that, many other server implementations are adopting them as well:

[Directives from GraphQL learning site](https://graphql.org/learn/queries/#directives)
[New features in GraphQL: Batch, defer, stream, live, and subscribe](https://dev-blog.apollodata.com/new-features-in-graphql-batch-defer-stream-live-and-subscribe-7585d0c28b07)
[GraphQL and Live Queries: Rodrigo Muñoz | POST/CON 2018](https://www.youtube.com/watch?v=TUgYyWC22og)

**Comment by johnberzy-bazinga** [comment]
Implemented in #166
